### PR TITLE
Popup updates, Auto Token Generation and Code fixes

### PR DESCRIPTION
**Pull request by rohit07sprinklr**
This adds token form to the the extension. Users can now update, remove and automatically generate private token. This private token is stored in google chrome local storage. 
Changes:
Previously, the token were constant and users had to manually edit the extension source file. After this, users can automatically generate a new Token or update existing token.

### CODE REVIEW

File: src/extension/ajaxClient.ts [hunk 1]
```diff
@@ -1,9 +1,23 @@
-import { DOMAIN_NAME,PRIVATE_TOKEN } from './constants/environmentVariables';
+const currentURL = new URL(window.location);
+const ORIGIN = currentURL.origin;
+
+async function getPrivateToken(){
```
Comment (Aditya7298) [inline src/extension/ajaxClient.ts:4]: try to fetch only once.

Comment (Aditya7298) [reply src/extension/ajaxClient.ts:4]: log on console on faliure.

File: src/extension/content.ts [hunk 2]
```diff
@@ -101,6 +102,12 @@ function renderDescription() {
 
 function render() {
   const rootDiv = document.createElement("div");
+  rootDiv.style.display='flex';
```
Comment (Aditya7298) [inline src/extension/content.ts:104]: single object

File: src/extension/content.ts [hunk 3]
```diff
@@ -132,19 +139,41 @@ function enableButtons() {
     el.disabled = false;
   });
 }
+function wait(millis) {
+  return new Promise((res) => setTimeout(res, millis));
+}
+
+async function waitCompleteRebase(repoURLName,mergeRequestID){
```
Comment (Aditya7298) [inline src/extension/content.ts]: getIsRebaseCompleted

File: src/extension/content.ts [hunk 4]
```diff
@@ -132,19 +139,41 @@ function enableButtons() {
...
   referenceEl.classList.add("mr-widget-workflow");
-  referenceEl.parentElement.prepend(el);
+  referenceEl.append(el);
 
   renderMergeButton(sourceBranch,targetBranch);
   const mergeButton = document.getElementById('gitlab-cli-merge');
   disableButtons();
   mergeButton.classList.remove(GITLAB_CLI_BUTTON);
   renderRebaseButton(repoURLName, mergeRequestID);
+  if(rebaseINProgress==true){
```
Comment (Aditya7298) [inline src/extension/content.ts]: isRebaseInProgress

File: src/extension/public/index.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  </head>
+  <body>
+    <div class="app-header">
+      <h2>GitLab API Extension</h2>
+    </div>
+    <hr>
+    <form id="token-form">
+      <label for="private-token">Personal access token:</label>
+      <input id ="private-token" type="text">
+      <div id="status-successhandler">Successfuly Updated!</div>
```
Comment (Aditya7298) [inline src/extension/public/index.html]: status-success-handler

File: src/extension/public/index.html [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+     <div id="status-errorhandler">Update failed!</div>
+      <button type="submit" id="form-submit" class="btn btn-primary btn-sm">Update</button>  
+      <button id="form-cleartoken" class="btn btn-danger btn-sm" type="button">Remove Token</button>  
+    </form>
+    <button id="generate_cli_token" style="margin-left: 5px;"><img src = "link.svg"></img>Generate New Token</button>
```
Comment (Aditya7298) [inline src/extension/public/index.html]: one pattern

File: src/extension/public/popup.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
+const generateToken= document.getElementById('generate_cli_token');
+async function getCurrentTab() {
+    let queryOptions = { active: true, lastFocusedWindow: true };
```
Comment (Aditya7298) [inline src/extension/public/popup.js]: const

File: src/extension/public/popup.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+});
+generateToken.addEventListener('click',async ()=>{
+    const currentTab = await getCurrentTab();
+    const currentURL = new URL(currentTab.url);
+    const ORIGN = currentURL.origin;
+    chrome.storage.sync.set({ 'gitTokenCopy':true });
+    chrome.tabs.create({active: true, url: `${ORIGN}/-/profile/personal_access_tokens?create_gitlabcli_token=true`});
+})
+
+function updateToken(){
```
Comment (Aditya7298) [inline src/extension/public/popup.js]: renderTokenValue

File: src/extension/public/popup.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+const form = document.getElementById('token-form');
+form.addEventListener('submit',async (e)=>{
+    e.preventDefault();
+    const ele = form.elements['private-token'];
+    const gitToken = ele.value;
+    try{
+        chrome.storage.sync.set({ gitToken });
+        const successhandler= document.getElementById('status-successhandler');
+        successhandler.style.display = 'block';
+        setTimeout(()=>{
```
Comment (Aditya7298) [inline src/extension/public/popup.js]: remove

File: src/extension/public/popup.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            successhandler.style.display = 'none';
+        },500);
+    }catch{
+        const errorhandler= document.getElementById('status-errorhandler');
+        errorhandler.style.display = 'block';
+        setTimeout(()=>{
```
Comment (Aditya7298) [inline src/extension/public/popup.js]: remove

File: src/extension/public/popup.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            errorhandler.style.display = 'none';
+        },500);
+    }
+});
+
+document.getElementById('form-cleartoken').addEventListener('click',async (e)=>{
```
Comment (Aditya7298) [inline src/extension/public/popup.js]: constant

File: src/extension/public/style.css [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    padding: 2px;
+    font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
+  }
+  .app-header{
+    background-color: rgb(235, 235, 235);
+    border-radius: 5px;
+    margin: 5px;
+    text-align: center;
+  }
+  #status-successhandler{
```
Comment (Aditya7298) [inline src/extension/public/style.css]: make classes

File: src/extension/token.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
+const main = () => {
+    const windowURL = new URL(window.location);
+    if(windowURL.searchParams.toString()=='create_gitlabcli_token=true'){
+        const inputArray = Array.from(document.querySelectorAll('input'))
```
Comment (Aditya7298) [inline src/extension/token.ts]: break function

File: src/extension/token.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+        const scopes = inputArray.filter(input => input.name === 'personal_access_token[scopes][]')
+        scopes.forEach((element)=>{
+            element.setAttribute("checked",true);
+        })
+        
+        const tokenName = inputArray.filter(input => input.name === 'personal_access_token[name]');
```
Comment (Aditya7298) [inline src/extension/token.ts]: find

File: src/extension/token.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        tokenName.forEach((element)=>{
+            element.value = 'GitlabCLIToken';
+        })
+
+        const tokenExpireDate = inputArray.filter(input => input.name === 'personal_access_token[expires_at]');
+        const today = new Date();
+        const date = (today.getFullYear()+1)+'-'+("0" + (today.getMonth()+1)).slice(-2)+'-'+("0" + (today.getDate()+1)).slice(-2);
```
Comment (Aditya7298) [inline src/extension/token.ts]: +2

Comment (Aditya7298) [reply src/extension/token.ts]: template string

File: src/extension/token.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+        const submitButton = inputArray.filter(input => input.name === 'commit');
+        submitButton.forEach((element)=>{
+            element.classList.remove('disabled');
+            element.removeAttribute("disabled");
+            element.click();
+        })
+    }
+    else{
+        const token = document.querySelector(".created-personal-access-token-container");
```
Comment (Aditya7298) [inline src/extension/token.ts:33]: //TODO - Can find better solution.

File: src/extension/content.ts [hunk 17]
```diff
@@ -132,50 +138,80 @@ function enableButtons() {
...
+  }
+}
 
-  try{
-    await fetch(`http://localhost:4000/handshake?location=${window.location}`)
-    .then((r) => {
-    if (r.status === 200) {
+async function initialise(repoURLName,mergeRequestID,sourceBranch,targetBranch,isRebaseInProgress) {
+  let retryCounter = 1;
+  while(retryCounter<=5){
```
Comment (Aditya7298) [inline src/extension/content.ts]: wrong. what will happen on slow connections ?

File: src/extension/content.ts [hunk 18]
```diff
@@ -188,23 +215,45 @@ function getProjectInfo(pathName){
...
+}
+const main = () => {
+  log("init");
+  const pathName = window.location.pathname;
+  const projectInfo = getProjectInfo(pathName);
+  
+  const targetNode = document.querySelector('.issuable-discussion');
+  const config = { childList: true, subtree: true };
+
+  const callback = function(mutationList, observer) {
```
Comment (Aditya7298) [inline src/extension/content.ts:242]: will this work if the main function is called after the mr-widget-section has already been added to the dom ?

File: src/extension/content.ts [hunk 19]
```diff
@@ -188,23 +215,45 @@ function getProjectInfo(pathName){
...
+      }
+  }
+  };
+
+  const observer = new MutationObserver(callback);
+  observer.observe(targetNode, config);
 };
 
-main();
+window.addEventListener ("load", main, false);
```
Comment (Aditya7298) [inline src/extension/content.ts:257]: why can't we call main directly ?

Comment (rohit07sprinklr) [reply src/extension/content.ts:257]: Yes, It will not be needed now

File: src/extension/content.ts (hunk 18 above)
Comment (rohit07sprinklr) [reply src/extension/content.ts:242]: Yes, there is a issue. We have to check if section has already been loaded

**Pull request merged**